### PR TITLE
feat: Add fine-tuning capabilities for TabPFN models

### PR DESCRIPTION
**Pull request by noahho**
This commit introduces a major new feature: the ability to fine-tune pre-trained TabPFN models on user-provided datasets. This allows for adapting the model to specific data distributions, potentially improving performance beyond its zero-shot capabilities.

Special thanks to Tobias Leemann (@tleemann) for initiating the work on this fine-tuning feature.

A new two-stage fine-tuning API is introduced:
1.  **`get_preprocessed_datasets()`**: Takes raw data and returns a `torch.utils.data.Dataset` object. It handles the various preprocessing and augmentation steps, preparing the data for a training loop.
2.  **`fit_from_preprocessed()`**: Called inside a training loop, this method fits the model on a batch of preprocessed data.

The model's `forward()` method is now exposed for use within the training loop to get predictions and calculate loss for backpropagation.

New examples demonstrating this workflow have been added:
- `examples/finetune_classifier.py`
- `examples/finetune_regressor.py`

- A new `fit_mode="batched"` and a corresponding `InferenceEngineBatchedNoPreprocessing` have been introduced to handle batches of pre-processed data efficiently.
- A `differentiable_input` flag enables end-to-end differentiability through the preprocessing pipeline.
- `DatasetCollectionWithPreprocessing` and a `meta_dataset_collator` utility have been created to integrate with PyTorch's `DataLoader`.
- Common logic between the classifier and regressor has been refactored into `base.py` to reduce code duplication.

- Dependency generation scripts have been consolidated into a single, more robust `generate_dependencies.py` script.
- Added a `.github/PULL_REQUEST_TEMPLATE.md` to standardize contributions.
- The attention mechanism has been refactored into its own `tabpfn.model.attention` submodule.

### CODE REVIEW

File: src/tabpfn/model/memory.py [hunk 1]
```diff
@@ -170,6 +170,8 @@ def estimate_memory_of_one_batch(
         Returns:
             The estimated memory usage of a single batch.
         """
+        assert len(X.shape) in (2, 3), "X must be a 2D or 3D tensor"
+
```
Comment (noahho) [inline src/tabpfn/model/memory.py:174]: abc
